File: src_local/agents/Python/sentinel.py
from src_local.agents.base import BaseActivePersona
import logging
import time

logger = logging.getLogger(__name__)

class SentinelPersona(BaseActivePersona):
    """Core: PhD in Cyber Security 🛡️"""
    
    def __init__(self, project_root):
        super().__init__(project_root)
        self.name, self.emoji, self.role, self.stack = "Sentinel", "🛡️", "PhD Security Architect", "Python"

    def perform_audit(self) -> list:
        start_time = time.time()
        logger.info(f"[{self.name}] Analisando Segurança...")
        
        audit_rules = [
            {'regex': r"(?<!['\"_])eval\(", 'issue': 'RCE: Execução dinâmica detectada.', 'severity': 'critical', 'skip_in_tests': False}, # Validado via AST agora
            {'regex': r"(?<!['\"_])shell=True", 'issue': 'Injeção: Shell ativo detectado.', 'severity': 'critical'},
            {'regex': r"(?i)(api_key|secret|token)\s*=\s*['\"][A-Za-z0-9_\-]{20,}['\"]", 'issue': 'Segurança: Segredo hardcoded detectado.', 'severity': 'critical'}
        ]
        
        results = self.find_patterns(('.py',), audit_rules)
        self._log_performance(start_time, len(results))
        return results

    def _reason_about_objective(self, objective, file, content):
        kw1, kw2 = "eval", "shell=True"
        if (kw1 in content or kw2 in content) and "rules =" not in content:
            return f"Vulnerabilidade: O objetivo '{objective}' exige integridade. Em '{file}', falhas de injeção comprometem a soberania da 'Orquestração de Inteligência Artificial'."
        return None

    def get_system_prompt(self):
        return f"Você é o Dr. {self.name}, mestre em segurança."

File: src_local/agents/Rust/Strategic/sentinel.ts
import { BaseActivePersona } from "../../base.ts";
import type { AuditFinding, AuditRule, StrategicFinding, ProjectContext } from "../../base.ts";

/**
 * 🛡️ Sentinel Persona (Rust Stack) - HYBRID VERSION
 * Especialista em integração estratégica e soberania da infraestrutura nativa.
 */
export class SentinelPersona extends BaseActivePersona {
    constructor(projectRoot?: string) {
        super(projectRoot);
        this.id = "rust:strategic:sentinel";
        this.name = "Sentinel";
        this.emoji = "🛡️";
        this.role = "Sovereign Security Architect";
        this.phd_identity = "Native Infrastructure Integrity & Safe Binding";
        this.stack = "Rust";
    }

    override getAuditRules(): { extensions: string[]; rules: AuditRule[] } {
        return {
            extensions: ['.proto', '.rs', 'main.go'], // Monitorando a ponte conectora
            rules: [
                { regex: /unsafe\s*\{/, issue: 'Memory Security: Bloco unsafe detectado. Audite meticulosamente os ponteiros raw PhD.', severity: 'high' }
            ]
        };
    }

    public override async execute(context: ProjectContext): Promise<(AuditFinding | StrategicFinding)[]> {
        this.setContext(context);
        const findings: (AuditFinding | StrategicFinding)[] = await this.performAudit();
        
        // 🛡️ POWER UP: Strategic Awareness via Knowledge Graph
        if (this.hub) {
            console.log("🛡️ [Sentinel-Rust] Consultando o Grafo de Conhecimento Nativo...");
            const graph = await this.hub.getKnowledgeGraph("src_native/analyzer/src/main.rs", 3);
            const securityQuery = await this.hub.queryKnowledgeGraph("unsafe", "high");
            const reasoning = await this.hub.reason(`Analyze Rust core sovereignty given ${securityQuery.length} unsafe blocks and core graph connectivity.`);

            findings.push({
                file: "Rust Core", agent: this.name, role: this.role, emoji: this.emoji,
                issue: `Sovereign Integrity: Núcleo Rust validado. Raciocínio: ${reasoning}`,
                severity: "INFO", stack: "Rust", evidence: `KG Depth: ${graph.nodes.length} nodes`, match_count: 1,
                context: "Native Security"
            } as any);
        }

        return findings;
    }

    public override async performAudit(): Promise<AuditFinding[]> {
        this.startMetrics();
        const rules = this.getAuditRules();
        const results = await this.findPatterns(rules.extensions, rules.rules);
        this.endMetrics(results.length);
        return results;
    }

    public override reasonAboutObjective(objective: string, file: string, _content: string | Promise<string | null>): StrategicFinding | null {
        return {
            file,
            issue: `Sovereignty Check: ${objective}. Garantindo que a ponte entre Runtime e Rust seja hermética PhD.`,
            severity: "STRATEGIC",
            context: this.name
        };
    }

    public override selfDiagnostic(): any {
        return {
            status: "Soberano",
            score: 100,
            issues: [],
            details: "Sentinela FFI operando com vigilância PhD."
        };
    }

    public override getSystemPrompt(): string {
        return `Você é o ${this.name}, guardião da soberania e integridade da ponte entre o mundo gerenciado e o mundo nativo (Rust).`;
    }
}
